### ISSUE TITLE
Gradle AOT  plugin needs test task to pass to build native image

### ISSUE THREAD

**Issue by phejl** [spring-projects-experimental/spring-native#873]
For some reason the native image build options are picked up from `build/resources/aotTest` so this works:
`./gradlew clean nativeBuild`
while this does not:
`./gradlew clean nativeBuild -x :test`

**Comment by phejl** [comment]
Can be workarounded by adding
```
nativeBuild {
    classpath("$buildDir/generated/resources/aot")
...
}
```

**Comment by phejl** [comment]
Possible workaround is
```
nativeBuild {
    ...
    classpath("$buildDir/resources/aot", "$buildDir/classes/java/aot")
}
```

**Comment by sdeleuze** [comment]
There is something wrong on Gradle side by on my side I see issues when the Gradle daemon is used (it is the default). Could you please try again with `--no-daemon`?

On my side what I see with the daemon after a second run:
```
16:04 $ ./gradlew --stacktrace -x :test nativeBuild
> Task :generateAot FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':generateAot'.
> Could not generate spring.factories source code
* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':generateAot'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$3(ExecuteActionsTaskExecuter.java:186)
...
org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:411)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:398)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:391)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:377)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.RuntimeException: Cannot resolve org.springframework.nativex.GeneratedClassNotFoundExceptionFailureAnalyzer
        at org.springframework.core.type.classreading.DefaultTypeSystem.resolve(DefaultTypeSystem.java:57)
        at org.springframework.core.type.classreading.TypeSystem.resolveClass(TypeSystem.java:34)
        at org.springframework.aot.factories.SpringFactory.resolve(SpringFactory.java:48)
        at org.springframework.aot.factories.SpringFactoriesContributor.loadSpringFactories(SpringFactoriesContributor.java:103)
        at org.springframework.aot.factories.SpringFactoriesContributor.contribute(SpringFactoriesContributor.java:67)
        ... 125 more
Caused by: java.util.zip.ZipException: ZipFile invalid LOC header (bad signature)
        at org.springframework.asm.ClassReader.readStream(ClassReader.java:321)
        at org.springframework.asm.ClassReader.<init>(ClassReader.java:287)
        at org.springframework.core.type.classreading.DefaultClassDescriptorReader.getClassReader(DefaultClassDescriptorReader.java:44)
        at org.springframework.core.type.classreading.DefaultClassDescriptorReader.readDescriptor(DefaultClassDescriptorReader.java:37)
        at org.springframework.core.type.classreading.DefaultTypeDescriptor.<init>(DefaultTypeDescriptor.java:46)
        at org.springframework.core.type.classreading.DefaultTypeSystem.resolve(DefaultTypeSystem.java:54)
```

cc @bclozel

**Comment by sdeleuze** [comment]
I have also performed this potentially related fix d7a9e3eeaee1f76422dacf4a1a7b2b90e30d01ca.

**Comment by sdeleuze** [comment]
I can reproduce it with `-no-daemon`, and I can't reproduce on https://github.com/graalvm/native-build-tools/tree/master/examples/gradle so that's likely something on our side. @bclozel Coul you please have to look to this one?

**Comment by phejl** [comment]
Yeah, sorry for late response. Was reproducible even with --no-daemon.

**Comment by bclozel** [comment]
I've looked into this problem and I've found a major problem between our own `org.springframework.experimental.aot` plugin and the `org.graalvm.buildtools.native` plugin.

With its 0.9.1 version, `org.graalvm.buildtools.native` will use the main and test sourceset classpath.
On our side, our plugin is generating, compiling sources and configuring the project like this:
* test aot sources are compiled using the test sourceset compile classpath and adding the resulting classes to the runtime classpath
* main aot sources are compiled using the main sourceset runtime classpath and adding the resulting classes to the runtime classpath considered by the `bootRun` and `bootJar` tasks.

As of 0.9.2+, the `nativeBuild` behavior has evolved and is [now resolving transitive project artifacts for the runtime classpath](https://github.com/graalvm/native-build-tools/blob/d1695ef3dc25b6b8a11e66c42e0459fa94a07844/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java#L166).

I think we should solve this problem in a two-steps approach:
1. document the configuration workaround suggested here, `classpath("$buildDir/resources/aot", "$buildDir/classes/java/aot")`
2. Get in touch with the `org.graalvm.buildtools.native` maintainers and refine our approach in the aot plugin.

**Comment by darxriggs** [comment]
Shouldn't `classpath(processAotResources.outputs, aotClasses.outputs)` instead of `classpath("$buildDir/resources/aot", "$buildDir/classes/java/aot")` be used, so Gradle can automatically add the dependency to these AOT tasks?

Otherwise it is required to add this additionally:
```
tasks.named('nativeBuild') {
    dependsOn 'aotClasses'
}
```

**Comment by sdeleuze** [comment]
@darxriggs Thanks for the proposal, I will let Brian provide you a feedback.

@bclozel We may release a 0.10.3 with related fixes in sync with native build tools 0.9.3, so feel free to commit on `main` in addition to `0.11.x` if you find something suitable for both branches.
